share/man/man5/fstab.5
.\"	$NetBSD: fstab.5,v 1.17 1999/09/23 16:03:12 tron Exp $
.\"
.\" Copyright (c) 1980, 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)fstab.5	8.1 (Berkeley) 6/5/93
.\"
.Dd June 5, 1993
.Dt FSTAB 5
.Os
.Sh NAME
.Nm fstab
.Nd filesystem table for devices, types, and mount points
.Sh SYNOPSIS
.Fd #include <fstab.h>
.Sh DESCRIPTION
The file
.Nm
contains descriptive information about the various file systems.
.Nm
is only read by programs, and not written;
it is the duty of the system administrator to properly create
and maintain this file.
Each filesystem is described on a separate line;
fields on each line are separated by tabs or spaces. Lines beginning
with
.Dq #
are comments.
The order of records in
.Nm
is important because
.Xr fsck 8 ,
.Xr mount 8 ,
and
.Xr umount 8
sequentially iterate through
.Nm
doing their respective tasks.
.Pp
The first field,
.Pq Fa fs_spec ,
describes the block special device or
remote filesystem to be mounted.
For filesystems of type
.Em ffs ,
the special file name is the block special file name,
and not the character special file name.
If a program needs the character special file name,
the program must create it by appending a
.Dq r
after the
last
.Dq /
in the special file name.
.Pp
The second field,
.Pq Fa fs_file ,
describes the mount point for the filesystem.
For swap and dump partitions, this field should be specified as
.Dq none .
.Pp
The third field,
.Pq Fa fs_vfstype ,
describes the type of the filesystem.
The system currently supports these filesystems:
.Bl -tag -width indent -offset indent
.It Em adosfs
an
.Tn AmigaDOS
filesystem
.It Em cd9660
an
.Tn ISO
9660 CD-ROM filesystem
.It Em ext2fs
an implementation of the Linux
.Dq Second Extended File-system .
.It Em fdesc
an implementation of /dev/fd
.It Em ffs
a local
.Ux
filesystem
.It Em kernfs
various and sundry kernel statistics
.It Em lfs
a log-structured file-system.  Currently does not work.
.It Em mfs
a local memory-based
.Ux
filesystem
.It Em msdos
an
.Tn MS-DOS
.Dq FAT filesystem
.It Em nfs
a Sun Microsystems compatible
.Dq Network File System
.It Em null
a loop-back filesystem, allowing parts of the system to be viewed
elsewhere.
.It Em portal
a general filesystem interface, currently supports TCP and FS mounts.
.It Em procfs
a local filesystem of process information
.It Em swap
a disk partition to be used for swapping and paging
.It Em umap
a user and group re-mapping filesystem.
.It Em union
a translucent filesystem
.El
.Pp
The fourth field,
.Pq Fa fs_mntops ,
describes the mount options associated with the filesystem.
It is formatted as a comma separated list of options.
It contains at least the type of mount (see
.Fa fs_type
below) plus any additional options
appropriate to the filesystem type.
.Pp
The option
.Dq auto
can be used in the
.Dq noauto
form to cause
a file system not to be mounted automatically (with
.Dq mount -a
,
or system boot time).
.Pp
If the options
.Dq userquota
and/or
.Dq groupquota
are specified,
the filesystem is automatically processed by the
.Xr quotacheck 8
command, and user and/or group disk quotas are enabled with
.Xr quotaon 8 .
By default,
filesystem quotas are maintained in files named
.Pa quota.user
and
.Pa quota.group
which are located at the root of the associated filesystem.
These defaults may be overridden by putting an equal sign
and an alternative absolute pathname following the quota option.
Thus, if the user quota file for
.Pa /tmp
is stored in
.Pa /var/quotas/tmp.user ,
this location can be specified as:
.Bd -literal -offset indent
userquota=/var/quotas/tmp.user
.Ed
.Pp
The type of the mount is extracted from the
.Fa fs_mntops
field and stored separately in the
.Fa fs_type
field (it is not deleted from the
.Fa fs_mntops
field).
If
.Fa fs_type
is
.Dq rw
or
.Dq ro
then the filesystem whose name is given in the
.Fa fs_file
field is normally mounted read-write or read-only on the
specified special file.
If
.Fa fs_type
is
.Dq sw
or
.Dq dp
then the special file is made available as a piece of swap
or dump
space by the
.Xr swapctl 8
command towards the beginning of the system reboot procedure.
See
.Xr swapctl 8
for more information on configuring swap and dump devices.
The fields other than
.Fa fs_spec
and
.Fa fs_type
are unused.
If
.Fa fs_type
is specified as
.Dq xx
the entry is ignored.
This is useful to show disk partitions which are currently unused.
.Pp
The fifth field,
.Pq Fa fs_freq ,
is used for these filesystems by the
.Xr dump 8
command to determine which filesystems need to be dumped.
If the fifth field is not present, a value of zero is returned and
.Xr dump 8
will assume that the filesystem does not need to be dumped.
.Pp
The sixth field,
.Pq Fa fs_passno ,
is used by the
.Xr fsck 8
program to determine the order in which filesystem checks are done
at reboot time.
The root filesystem should be specified with a
.Fa fs_passno
of 1, and other filesystems should have a
.Fa fs_passno
of 2.
Filesystems within a drive will be checked sequentially,
but filesystems on different drives will be checked at the
same time to utilize parallelism available in the hardware.
If the sixth field is not present or zero,
a value of zero is returned and
.Xr fsck 8
will assume that the filesystem does not need to be checked.
.Bd -literal
#define	FSTAB_RW	"rw"	/* read-write device */
#define	FSTAB_RO	"ro"	/* read-only device */
#define	FSTAB_SW	"sw"	/* swap device */
#define	FSTAB_DP	"dp"	/* dump device */
#define	FSTAB_XX	"xx"	/* ignore totally */

struct fstab {
	char	*fs_spec;	/* block special device name */
	char	*fs_file;	/* filesystem path prefix */
	char	*fs_vfstype;	/* type of filesystem */
	char	*fs_mntops;	/* comma separated mount options */
	char	*fs_type;	/* rw, ro, sw, or xx */
	int	fs_freq;	/* dump frequency, in days */
	int	fs_passno;	/* pass number on parallel fsck */
};
.Ed
.Pp
The proper way to read records from
.Pa fstab
is to use the routines
.Xr getfsent 3 ,
.Xr getfsspec 3 ,
.Xr getfstype 3 ,
and
.Xr getfsfile 3 .
.Sh FILES
.Bl -tag -width /etc/fstab -compact
.It Pa /etc/fstab
The file
.Nm
resides in
.Pa /etc .
.El
.Sh SEE ALSO
.Xr getfsent 3 ,
.Xr mount_ados 8 ,
.Xr mount_cd9660 8 ,
.Xr mount_ext2fs 8 ,
.Xr mount_fdesc 8 ,
.Xr mount_ffs 8 ,
.Xr mount_kernfs 8 ,
.Xr mount_lfs 8 ,
.Xr mount_mfs 8 ,
.Xr mount_msdos 8 ,
.Xr mount_nfs 8 ,
.Xr mount_null 8 ,
.Xr mount_portal 8 ,
.Xr mount_procfs 8 ,
.Xr mount_ufs 8 ,
.Xr mount_umap 8 ,
.Xr mount_union 8 ,
.Xr swapctl 8 .
.Sh HISTORY
The
.Nm
file format appeared in
.Bx 4.0 .


lib/libc/gen/fstab.c
/*	$NetBSD: fstab.c,v 1.22 2000/01/22 22:19:10 mycroft Exp $	*/

/*
 * Copyright (c) 1980, 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(LIBC_SCCS) && !defined(lint)
#if 0
static char sccsid[] = "@(#)fstab.c	8.1 (Berkeley) 6/4/93";
#else
__RCSID("$NetBSD: fstab.c,v 1.22 2000/01/22 22:19:10 mycroft Exp $");
#endif
#endif /* LIBC_SCCS and not lint */

#include "namespace.h"
#include <sys/types.h>

#include <assert.h>
#include <err.h>
#include <errno.h>
#include <fstab.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifdef __weak_alias
__weak_alias(endfsent,_endfsent)
__weak_alias(getfsent,_getfsent)
__weak_alias(getfsfile,_getfsfile)
__weak_alias(getfsspec,_getfsspec)
__weak_alias(setfsent,_setfsent)
#endif

static FILE *_fs_fp;
static size_t _fs_lineno = 0;
static const char *_fs_file = _PATH_FSTAB;
static struct fstab _fs_fstab;

static int fstabscan __P((void));

static __inline char *nextfld __P((char **, const char *));


static __inline char *
nextfld(str, sep)
	char **str;
	const char *sep;
{
	char *ret;

	_DIAGASSERT(str != NULL);
	_DIAGASSERT(sep != NULL);

	while ((ret = strsep(str, sep)) != NULL && *ret == '\0')
		continue;
	return ret;
}


static int
fstabscan()
{
	char *cp, *lp, *sp;
#define	MAXLINELENGTH	1024
	static char line[MAXLINELENGTH];
	char subline[MAXLINELENGTH];
	static const char sep[] = ":\n";
	static const char ws[] = " \t\n";
	static char *fstab_type[] = {
	    FSTAB_RW, FSTAB_RQ, FSTAB_RO, FSTAB_SW, FSTAB_DP, FSTAB_XX, NULL
	};

	for (;;) {
		if (!(lp = fgets(line, sizeof(line), _fs_fp)))
			return 0;
		_fs_lineno++;
/* OLD_STYLE_FSTAB */
		if (!strpbrk(lp, " \t")) {
			_fs_fstab.fs_spec = nextfld(&lp, sep);
			if (!_fs_fstab.fs_spec || *_fs_fstab.fs_spec == '#')
				continue;
			_fs_fstab.fs_file = nextfld(&lp, sep);
			_fs_fstab.fs_type = nextfld(&lp, sep);
			if (_fs_fstab.fs_type) {
				if (!strcmp(_fs_fstab.fs_type, FSTAB_XX))
					continue;
				_fs_fstab.fs_mntops = _fs_fstab.fs_type;
				_fs_fstab.fs_vfstype =
				    strcmp(_fs_fstab.fs_type, FSTAB_SW) ?
				    "ufs" : "swap";
				if ((cp = nextfld(&lp, sep)) != NULL) {
					_fs_fstab.fs_freq = atoi(cp);
					if ((cp = nextfld(&lp, sep)) != NULL) {
						_fs_fstab.fs_passno = atoi(cp);
						return 1;
					}
				}
			}
			goto bad;
		}
/* OLD_STYLE_FSTAB */
		_fs_fstab.fs_spec = nextfld(&lp, ws);
		if (!_fs_fstab.fs_spec || *_fs_fstab.fs_spec == '#')
			continue;
		_fs_fstab.fs_file = nextfld(&lp, ws);
		_fs_fstab.fs_vfstype = nextfld(&lp, ws);
		_fs_fstab.fs_mntops = nextfld(&lp, ws);
		if (_fs_fstab.fs_mntops == NULL)
			goto bad;
		_fs_fstab.fs_freq = 0;
		_fs_fstab.fs_passno = 0;
		if ((cp = nextfld(&lp, ws)) != NULL) {
			_fs_fstab.fs_freq = atoi(cp);
			if ((cp = nextfld(&lp, ws)) != NULL)
				_fs_fstab.fs_passno = atoi(cp);
		}
		sp = strncpy(subline, _fs_fstab.fs_mntops, sizeof(subline)-1);
		while ((cp = nextfld(&sp, ",")) != NULL) {
			char **tp;

			if (strlen(cp) != 2)
				continue;

			for (tp = fstab_type; *tp; tp++)
				if (strcmp(cp, *tp) == 0) {
					_fs_fstab.fs_type = *tp;
					break;
				}
			if (*tp)
				break;
		}
		if (strcmp(_fs_fstab.fs_type, FSTAB_XX) == 0)
			continue;
		if (cp != NULL)
			return 1;

bad:
		warnx("%s, %lu: Missing fields", _fs_file, (u_long)_fs_lineno);
	}
	/* NOTREACHED */
}

struct fstab *
getfsent()
{
	if ((!_fs_fp && !setfsent()) || !fstabscan())
		return NULL;
	return &_fs_fstab;
}

struct fstab *
getfsspec(name)
	const char *name;
{

	_DIAGASSERT(name != NULL);

	if (setfsent())
		while (fstabscan())
			if (!strcmp(_fs_fstab.fs_spec, name))
				return &_fs_fstab;
	return NULL;
}

struct fstab *
getfsfile(name)
	const char *name;
{

	_DIAGASSERT(name != NULL);

	if (setfsent())
		while (fstabscan())
			if (!strcmp(_fs_fstab.fs_file, name))
				return &_fs_fstab;
	return NULL;
}

int
setfsent()
{
	_fs_lineno = 0;
	if (_fs_fp) {
		rewind(_fs_fp);
		return 1;
	}
	if ((_fs_fp = fopen(_PATH_FSTAB, "r")) == NULL) {
		warn("Cannot open `%s'", _PATH_FSTAB);
		return 0;
	}
	return 1;
}

void
endfsent()
{
	if (_fs_fp) {
		(void)fclose(_fs_fp);
		_fs_fp = NULL;
	}
}
